docs/src/nondiff_points.md

```diff
@@ -0,0 +1,141 @@
+    If it is not differentiable due to the primal not being defined on one side, you can set it to what ever you like.
+    Your rule should claim a derivative that is *useful*.
+In calculus one learns that if the derivative as computed by approaching from the left,
+and the derivative one computes as approaching from the right are not equal then the derivative is not defined,
+and we say the function is not differentiable at that point.
+This is distinct from the notion captured by [`NoTangent`](@ref), which is that the tangent space itself is not defined: because in some sense the primal value can not be perturbed e.g. is is a discrete type.
+
+However, contrary to what calculus says most autodiff systems will return an answer for such functions.
+For example for: `abs_left(x) = (x <= 0) ? -x : x`, AD will say the derivative at `x=0` is `-1`.
+Alternatively for:  `abs_right(x) = (x < 0) ? -x : x`, AD will say the derivative at `x=0` is `1`.
+Those two examples are weird since they are equal at all points, but AD claims different derivatives at `x=0`.
+The way to fix autodiff systems being weird is to write custom rules.
+So what rule should we write for this case?
+
+The obvious answer, would be to write a rule that throws an error if input at a point where calculus says the derivative is not defined.
+Another option is to return some error signally value like `NaN`.
+Which you *can* do.
+However, there is no where to go with an error, the user still wants a derivative; so this is not useful.
+
+Let us explore what is useful:
+## Case Studies
+
+```@setup nondiff
+using Plots
+gr(framestyle=:origin, legend=false)
+```
+### Derivative is defined in usual sense
+```@example nondiff
+plot(x->x^3)
+```
+This is the standard case, one can returned the derivative that is defined according to school room calculus.
+Here we would reasonably say that at `x=0` the derivative is `3*0^2=0`. 
+
+
+An interesting thing about `x->x^3` is that at `x=0` the derivative is defined, but neither the sub-derivative nor super-derivative is defined.
+
+### Local Minima / Maxima
+
+```@example nondiff
+plot(abs)
+```
+
+`abs` is the classic example of a function where the derivative is not defines as the limit from above is not equal to the limit from below
+
+$$\mathrm{abs}'(0) = \lim_{h->0^-} \dfrac{\mathrm{abs}(0)-\mathrm{abs}(0-h)}{0-h} = -1$$
+$$\mathrm{abs}'(0) = \lim_{h->0^+} \dfrac{abs(0)-\mathrm{abs}(0-h)}{0-h} = 1$$
+
+Now, as discussed in the introduction the AD system would on it's own choose either 1 or -1, depending on implementation.
+
+We however have a potentially much nicer answer available to use: 0.
+
+This has a number of advantages.
+- It follows the rule that derivatives are zero at local minima (and maxima).
+- If you leave a gradient decent optimizer running it will eventually actually converge absolutely to the point -- where as with it being 1 or -1 it would never outright converge it would always flee.
```

Comment:
```suggestion
- If you leave a gradient decent optimizer running it will eventually actually converge absolutely to the point -- where as with it being 1 or -1 it would never outright converge it would always flee.
```
The word "flee" is evocative, but maybe a little confusing here. Maybe instead we could say "oscillate" or "wobble"